tinylfu.go

```diff
@@ -5,55 +5,119 @@
 package tinylfu
 
 import (
-	"container/list"
 	"github.com/dgryski/go-metro"
 )
 
-type T struct {
-	c       *cm4
-	bouncer *doorkeeper
-	w       int
-	samples int
-	lru     *lruCache
-	slru    *slruCache
-	data    map[string]*list.Element
+type T[V any] struct {
+	c              *cm4
+	bouncer        *doorkeeper
+	w              int
+	samples        int
+	lru            *lruCache[V]
+	slru           *slruCache[V]
+	data           map[string]*Element[slruItem[V]]
+	hits           uint64
+	misses         uint64
+	lruPct         float32
+	interval       int
+	step           int
+	percentage     float32
+	wentUp         bool
+	madeStepBigger bool
+	lastSuccess    float32
+	size           int
+	withResize     bool
 }
 
-func New(size int, samples int) *T {
+func New[V any](size int, samples int, withResize bool) *T[V] {
 
 	const lruPct = 1
-
 	lruSize := (lruPct * size) / 100
 	if lruSize < 1 {
 		lruSize = 1
 	}
 	slruSize := int(float64(size) * ((100.0 - lruPct) / 100.0))
 	if slruSize < 1 {
-		slruSize  = 1
+		slruSize = 1
 
 	}
 	slru20 := int(0.2 * float64(slruSize))
 	if slru20 < 1 {
 		slru20 = 1
 	}
 
-	data := make(map[string]*list.Element, size)
+	data := make(map[string]*Element[slruItem[V]], size)
 
-	return &T{
+	return &T[V]{
 		c:       newCM4(size),
 		w:       0,
 		samples: samples,
 		bouncer: newDoorkeeper(samples, 0.01),
 
 		data: data,
 
-		lru:  newLRU(lruSize, data),
-		slru: newSLRU(slru20, slruSize-slru20, data),
+		lru:  newLRU[V](lruSize, data),
+		slru: newSLRU[V](slru20, slruSize-slru20, data),
+
+		percentage: 6.25,
+		size:       size,
+		lruPct:     lruPct,
+		step:       50000,
+
+		withResize: withResize,
 	}
 }
 
-func (t *T) Get(key string) (interface{}, bool) {
+func (t *T[V]) resize() {
```

Comment:
There should be more docs around the resize logic and the associated configuration parameters.

Reply:
i see now that the logic i implemented is incorrect, as i am not conserving the order of the elements.
i will redo.